nucypher/characters/control/controllers.py

```diff
@@ -164,42 +167,225 @@ def __init__(self,
 
         super().__init__(control_serializer=self._internal_controller.serializer)
 
         self.log = Logger(app_name)
 
-    def make_web_controller(self):
+    @abstractmethod
+    def make_control_transport(self):
+        return NotImplemented
+
+    @abstractmethod
+    def handle_request(self, interface, control_request, *args, **kwargs):
+        return NotImplemented
+
+    @abstractmethod
+    def test_client(self):
+        return NotImplemented
+
+
+class JSONRPCController(CharacterControlServer):
+
+    _emitter_class = JSONRPCStdoutEmitter
+
+    def start(self):
+        _transport = self.make_control_transport()
+        reactor.run()  # < ------ Blocking Call (Reactor)
+
+    def test_client(self) -> JSONRPCTestClient:
+        test_client = JSONRPCTestClient(rpc_controller=self)
+        return test_client
+
+    def make_control_transport(self):
+        transport = stdio.StandardIO(JSONRPCLineReceiver(rpc_controller=self))
+        return transport
+
+    def get_interface(self, name: str) -> Callable:
+
+        # Examine the controller interface
+        interfaces = inspect.getmembers(self._internal_controller,
+                                        predicate=inspect.ismethod)
+
+        # Generate a mapping of the public interfaces
+        interfaces = {i[0]: i[1] for i in interfaces if not i[0].startswith('_')}
+
+        try:
+            interface = interfaces[name]
+        except KeyError:
+            raise self.emitter.MethodNotFound
+
+        return interface
+
+    def handle_server_notification(self, notification_request) -> int:
+        pass
+
+    def handle_procedure_call(self, control_request, *args, **kwargs) -> int:
+
+        # Validate request and read request metadata
+        jsonrp2 = control_request['jsonrpc']
+        if jsonrp2 != '2.0':
+            raise self.emitter.InvalidRequest
+
+        request_id = control_request['id']
+
+        # Read the interface's signature metadata
+        method_name = control_request['method']
+        method_params = control_request.get('params', dict())  # optional
+
+        # Lookup the public interface
+        interface = self.get_interface(name=method_name)
+
+        # Call the internal interface | pipe to output
+        return interface(request=method_params, request_id=request_id, *args, **kwargs)  # < ------- INLET
+
+    def validate_request(self, request: dict):
+
+        #
+        # Phase 1 - Metadata
+        #
+
+        required_fileds = {'jsonrpc', 'method'}
+        optional_fields = {'id', 'params'}
+        all_fields = required_fileds | optional_fields
+
+        try:
+            input_fields = set(request.keys())
+        except AttributeError:
+            raise self.emitter.InvalidRequest
+
+        contains_required_fields = required_fileds.issubset(input_fields)
+
+        uniquie_fields = all_fields - input_fields - optional_fields
+        contains_valid_fields = not bool(uniquie_fields)
```

Comment:
I think this logic is unnecessarily complex and I don't see immediately what it's trying to achieve. 

Since `required_fileds` and `optional_fields` are disjoint sets (otherwise their names would be inconsistent) and `all_fields = required_fileds | optional_fields`, then the expression `all_fields - input_fields - optional_fields` is always equal to `required_fields - input_fields`.

If `contains_required_fields` is True (i.e., `required_fileds` is a subset of `input_fields`), then the latter expression is always the empty set, so `contains_valid_fields` is True too; on the contrary, if `contains_required_fields` is False, then the expression is not empty and `contains_valid_fields` is False. 

Hence, it's always the case that `contains_required_fields == contains_valid_fields`, so we can get rid of the `contains_valid_fields` check.

Also: `uniquie` --> `unique`

Reply:
Here, I am detecting the possibility of an unknown field being present in `all_fields`.

`unique_fields = all_fields - input_fields - optional_fields` - This leaves me with a set containing only the unknown fields.